### ISSUE TITLE
Arbitrary metadata for uproot transformers

### ISSUE THREAD

**Issue by BenGalewsky** [ssl-hep/ServiceX#643]
_As an analyzer I want to be able to provide arbitrary metadata for the uproot transformers so I can encode more complicated behaviors

# Description
Initially this makes more sense with the python code generator, but it's possible this could be picked up by the uproot func_adl transformer.

We would add a new method to the `Dataset` class, `with_metadata`. Here's how it would look in practice:
```python
ds = sx.python_dataset(dataset_id, codegen="python", title="Python")
def run_query(input_filenames=None, metadata):
    import uproot
    o = uproot.lazy({input_filenames: "mini"})
    x = metadata['special_value']
    return o.lep_pt


meta = {"special_value": 42}
sx3 = ds.with_uproot_function(run_query).with_metadata(meta).as_pandas()
```

**Comment by gordonwatts** [comment]
I can't tell how the metadata is being used here. The user story doesn't tell me enough about "more complex behaviors".

**Comment by alexander-held** [comment]
This is meant to enable behavior in the Python function that depends on metadata (like sample-dependent choices to take). Such an interface enables this in a cleaner manner than relying on filename parsing within the function.

Random example: when handling ttbar and tt+bb samples, take care of overlap removal by applying suitable cuts.

**Comment by gordonwatts** [comment]
Ah - ok - so this is just a change in the user interface, the backend wouldn't be affected.

So - yes - that makes sense to me. `coffea` uses this approach quite a bit, actually, as a way to punch through multiple layers  (shades of dealing with `ignore_cache` here, @BenGalewsky ).

**Comment by alexander-held** [comment]
What I have in mind still affects the backend: the metadata would be transported through to the transformer executing the Python function, so that within that function `metadata` is accessible.

**Comment by gordonwatts** [comment]
Ok - now I understand better. First, the obvious, a `DatasetGroup` is so that you can treat all the datasets the same. The `metadata` feature here is so you can treat members of a `DataSetGroup` differently. You are pushing down the `if` statements based on dataset into the code that runs parallel, rather than keeping it at the top level in the "steering" code. I think we mostly due this because this is how we've always written physics code - one giant monolithic block of code with if statements, rather than smaller composable bits we put together in different orders depending. Both work, obviously!

Looking over to `func_adl` there is totally some ways I can see using this. For example, I can see wanting to have the same sequence of data manipulations, but with slightly different cuts depending on the region (control, signal, etc.) we are looking at.

For implementation - I think we should think about if the backend should be changed at all. I'd much prefer to leave this complexity only in the front end. For example, the python code transmitted down could be turned into:

```python
metadata = {
  "name": "ttbar"
  "do_ttbar_filtering": False
}

def processing_function(data):
  if metadata["do_ttbar_filtering"]:
    do-it-one-way()
  else:
    do-it-the-other()
```

For implementation - for `func_adl` this feels like just another form of variable/global capture. The code that lifts `lambda`'s out in `func_adl` already does lambda capture. The `metadata` is already available in the code before it goes to `servicex`. Which makes me ask... Should this occur in the front end or in the backend? Feels more natural to leave ServiceX out of it.

BTW, are globals captured for the code before it is sent down for the python transformer? When a function is defined inline this is super helpful:

```python

def do_the_work(ds, parameter):

    def run_query(input_filenames=None, metadata):
        import uproot
        o = uproot.lazy({input_filenames: "mini"})
        x = metadata['special_value']
        return o.lep_pt + parameter

    meta = {"special_value": 42}
    sx3 = ds.with_uproot_function(run_query).with_metadata(meta).as_pandas()

do_the_work(ds, 20.0)
```

One could use the same mechanism here.

Another thing is that the metadata should be available by querying the "dataset" at any point.

To Summarize, here are my thoughts:

* The dataset ctor can add metadata.
* One can retrieve the metadata at any time one wants by looking at the `dataset.metadata` (or whatever it ends up being called).
* The `metadata` is available in both `func_adl` and for the python downstream
* Figure out how to do this without modifying the backend if at all possible.
* For a dataset group the `.metadata` doesn't make sense, though perhaps we could have an apply function (??) that might be fun, like `dsg_new = dsg.apply(lambda ds: ds.metadata.do_filtering ? ds.Where(lambda evt: filter(evt)) : ds)`

**Comment by BenGalewsky** [comment]
This should be implemented as a per-codegen solution
